docs/dev/Pagination-v2.md

```diff
@@ -100,10 +111,381 @@ The discussion below uses *under max_result_window* to refer to scenarios that c
 ## SQL Node Load Balancing
 V2 SQL engine supports *sql node load balancing* -- a cursor request can be routed to any SQL node in a cluster. This is achieved by encoding all data necessary to retrieve the next page in the `cursor_id`.
 
-## Design Diagrams
-New code workflows are highlighted.
+## Feature Design
+
+### Plan Tree changes
+
+There are different plan trees are built during request processing. Read more about their purpose and stages [here](query-optimizer-improvement.md#Examples). Article below describes what changes are introduced in these trees by pagination feature.
+
+Simplified workflow of plan trees is shown below. Initial Page Request is processed the same way as non-paging request.
+
+```mermaid
+stateDiagram-v2
+  state "Non Paged Request" as NonPaged {
+    direction LR
+    state "Parse Tree" as Parse
+    state "Unresolved Plan Tree" as Unresolved
+    state "Abstract Plan Tree" as Abstract
+    state "Logical Plan Tree" as Logical
+    state "Optimized Logical Plan Tree" as Optimized
+    state "Physical Plan Tree" as Physical
+
+    [*] --> Parse : ANTLR
+    Parse --> Unresolved : AstBuilder
+    Unresolved --> Abstract : QueryPlanner
+    Abstract --> Logical : Planner
+    Logical --> Optimized : Optimizer
+    Optimized --> Physical : Implementor
+  }
+```
+```mermaid
+stateDiagram-v2
+  state "Initial Page Request" as Paged {
+    direction LR
+    state "Parse Tree" as Parse
+    state "Unresolved Plan Tree" as Unresolved
+    state "Abstract Plan Tree" as Abstract
+    state "Logical Plan Tree" as Logical
+    state "Optimized Logical Plan Tree" as Optimized
+    state "Physical Plan Tree" as Physical
+
+    [*] --> Parse : ANTLR
+    Parse --> Unresolved : AstBuilder
+    Unresolved --> Abstract : QueryPlanner
+    Abstract --> Logical : Planner
+    Logical --> Optimized : Optimizer
+    Optimized --> Physical : Implementor
+  }
+```
+```mermaid
+stateDiagram-v2
+  state "Subsequent Page Request" as Paged {
+    direction LR
+    state "Abstract Plan Tree" as Abstract
+    state "Physical Plan Tree" as Physical
+
+    [*] --> Abstract : QueryPlanner
+    Abstract --> Physical : Deserializer
+  }
+```
+
+New plan tree workflow was added for Subsequent Page Requests. Since a final Physical Plan tree was already created for Initial request, subsequent requests should have the same tree. The tree is being serialized into `cursor` by `PlanSerializer` and then restored back. Query parsing and analysis is also not performed for cursor requests, so these steps are also not executed there.
```

Comment:
```suggestion
New plan tree workflow was added for Subsequent Page Requests. Since a final Physical Plan tree was already created for Initial request, subsequent requests should have the same tree. The tree is serialized into a `cursor` by `PlanSerializer` to be de-serialized on the subsequence Page Request. Query parsing and analysis is not performed for Subsequent Page Requests, since the Physical Plan tree is instead de-serialized from the `cursor`.
```